### PR TITLE
[ZEPPELIN-1164] ZeppelinHub Realm

### PR DESCRIPTION
**Pull request by anthonycorbacho**
### What is this PR for?
Add a new Shiro Realm that lets users authenticate in Apache Zeppelin using their ZeppelinHub credentials.

### What type of PR is it?
`Feature`

### Todos
* [x] - Create ZeppelinHub Realm that extends `AuthorizingRealm`.
* [x] - Update `shiro.ini` file and add ZeppelinHub configuration template. 
* [x] - Add documentation

### What is the Jira issue?
[Issue 1164](https://issues.apache.org/jira/browse/ZEPPELIN-1164)

### How should this be tested?
 - Assuming that you already have an account [Zeppelinhub](https://www.zeppelinhub.com/register), edit `conf/shiro.ini` file and uncomment ZeppelinHub related configuration.
 - Start your Zeppelin
 - Login with your ZeppelinHub credentials.

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? YES

### CODE REVIEW

File: zeppelin-server/src/main/java/org/apache/zeppelin/realm/ZeppelinHubRealm.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  private final HttpClient httpClient;
+  private final Gson gson;
+
+  private String zeppelinhubUrl;
+  private String name;
+
+  public ZeppelinHubRealm() {
+    super();
+    LOG.debug("Init ZeppelinhubRealm");
+    //TODO(anthonyc): think about more setting for this HTTP client. eg: if user use proxy etcetc...
```
Comment (khalidhuseynov) [inline zeppelin-server/src/main/java/org/apache/zeppelin/realm/ZeppelinHubRealm.java]: may need to address line lenght here, build fails with 
```
ZeppelinHubRealm.java[68] (sizes) LineLength: Line is longer than 100 characters (found 101).
```

**Comment by khalidhuseynov** [comment]
build failed on `zeppelin-web`, which is not relevant. can we still re-trigger CI here?

**Comment by anthonycorbacho** [comment]
all green

File: zeppelin-server/src/main/java/org/apache/zeppelin/realm/ZeppelinHubRealm.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  private String name;
+
+  public ZeppelinHubRealm() {
+    super();
+    LOG.debug("Init ZeppelinhubRealm");
+    //TODO(anthonyc): think about more setting for this HTTP client.
+    //                eg: if user uses proxy etcetc...
+    httpClient = new HttpClient();
+    gson = new Gson();
+    name = getClass().getName() + "_" + INSTANCE_COUNT.getAndIncrement();
```
Comment (khalidhuseynov) [inline zeppelin-server/src/main/java/org/apache/zeppelin/realm/ZeppelinHubRealm.java]: one question, why we have `INSTANCE_COUNT` here and how `name` is used?

Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/org/apache/zeppelin/realm/ZeppelinHubRealm.java]: its a shiro specific needs for unique name.

**Comment by khalidhuseynov** [comment]
It looks good. one thing is that it should be documented as well so that others may implement their own authentication/authorization realms with their server. do you want to add it as a part of this PR or do it later?

**Comment by anthonycorbacho** [comment]
@khalidhuseynov i think it make sense to do it in this PR.

**Comment by anthonycorbacho** [comment]
@khalidhuseynov @AhyoungRyu 5a27871f5a9e436653c8862e6d2ecc8913fdd496 for documentation

File: docs/security/shiroauthentication.md [hunk 3]
```diff
@@ -121,5 +148,5 @@ Since Shiro provides **url-based security**, you can hide the information by com
 In this case, only who have `admin` role can see **Interpreter Setting**, **Credential** and **Configuration** information. 
 If you want to grant this permission to other users, you can change **roles[ ]** as you defined at `[users]` section.
 
-<br/>
-> **NOTE :** All of the above configurations are defined in the `conf/shiro.ini` file. This documentation is originally from [SECURITY-README.md](https://github.com/apache/zeppelin/blob/master/SECURITY-README.md).
```
Comment (AhyoungRyu) [inline docs/security/shiroauthentication.md]: Is there any particular reason to remove this sentence?

"All of the above configurations are defined in the `conf/shiro.ini` file."

Comment (anthonycorbacho) [inline docs/security/shiroauthentication.md]: IDK its after doing my rebase. should i bring it back?

Comment (khalidhuseynov) [inline docs/security/shiroauthentication.md]: i think it exists in master, so may need to bring it back. @AhyoungRyu as original author may check it

Comment (anthonycorbacho) [inline docs/security/shiroauthentication.md]: Just pushed the fix, @AhyoungRyu sorry about that.

**Comment by khalidhuseynov** [comment]
LGTM

**Comment by AhyoungRyu** [comment]
LGTM 👍

File: docs/security/shiroauthentication.md (hunk 3 above)
Comment (AhyoungRyu) [inline docs/security/shiroauthentication.md]: @anthonycorbacho It's fine :) @khalidhuseynov thanks for your explanation!

**Comment by anthonycorbacho** [comment]
merging if no more discussion

**Comment by jongyoul** [comment]
@anthonycorbacho I've tried to merge this PR into master and branch-0.6. I've succeeded to merge it into master but failed doing into branch-0.6. Is it OK?

**Comment by anthonycorbacho** [comment]
@jongyoul sure, no problemo
Thanks

**Pull request closed without merging**